src/scheduler/determinism_check.rs

```diff
@@ -0,0 +1,141 @@
+use std::iter::FromIterator;
+
+use crate::runtime::task::TaskId;
+use crate::scheduler::{Schedule, ScheduleStep, Scheduler};
+
+/// A `ScheduleRecord` can be used to record both the step
+/// taken and all runnable tasks
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct ScheduleRecord {
+    step: ScheduleStep,
+    runnable_tasks: HashSet<TaskId>,
+}
+
+impl ScheduleRecord {
+    /// Create a record step from the chosen task and options
+    fn new(step: ScheduleStep, runnable_tasks: &[TaskId]) -> Self {
+        let runnable_tasks = runnable_tasks.iter().cloned().collect::<HashSet<_>>();
+        Self { step, runnable_tasks }
+    }
+}
+
+/// /// A `DeterminismCheckScheduler` checks whether a given program is deterministic
+/// by wrapping an inner `Scheduler` and, for each schedule generated by that scheduler,
+/// replaying the schedule a chosen number of times. On each replay, we check that the
+/// schedule is still valid and that the set of runnable tasks is the same at each step.
+/// Violations of checks checks mean that the program is not deterministic.
+#[derive(Debug)]
+pub struct DeterminismCheckScheduler<S: ?Sized + Scheduler> {
+    inner: Box<S>,
+    iterations: usize,
+    inner_iterations: usize,
+    recording: bool,
+    original_schedule: Vec<ScheduleRecord>,
+    current_step: usize,
+}
+
+impl<S: Scheduler> DeterminismCheckScheduler<S> {
+    /// Create a new `DeterminismCheckScheduler` by wrapping the given `Scheduler` implementation.
+    pub fn new(inner_iterations: usize, inner: S) -> Self {
+        Self {
+            inner: Box::new(inner),
+            iterations: 0,
+            inner_iterations,
+            original_schedule: Vec::new(),
+            recording: false,
+            current_step: 0,
+        }
+    }
+}
+
+impl<S: Scheduler> Scheduler for DeterminismCheckScheduler<S> {
+    fn new_execution(&mut self) -> Option<Schedule> {
+        if self.iterations % self.inner_iterations == 0 {
+            // Start a new recording
+            self.recording = true;
+            self.original_schedule.clear();
+        } else {
+            // Create a new execution to test against prior recording
+
+            if self.recording {
+                self.recording = false;
+            }
+
+            if self.current_step != self.original_schedule.len() {
+                panic!("possible nondeterminism: current execution ended earlier than expected (expected length {} but ended after {})", self.original_schedule.len(), self.current_step);
+            }
+        }
+
+        self.current_step = 0;
+        self.iterations += 1;
+
+        self.inner.new_execution()
+    }
+
+    fn next_task(
+        &mut self,
+        runnable_tasks: &[TaskId],
+        current_task: Option<TaskId>,
+        is_yielding: bool,
+    ) -> Option<TaskId> {
+        if self.recording {
+            // Recording a schedule
+            let choice = self.inner.next_task(runnable_tasks, current_task, is_yielding).unwrap();
```

Comment:
Not sure unwrap is the right thing here? The inner scheduler is allowed to return `None` to stop early, and we should respect that here. (This might also be worth a unit test -- use an inner scheduler that stops after 5 steps or something).